crates/autopilot/src/on_settlement_event_updater.rs

```diff
@@ -129,208 +114,60 @@ impl Inner {
             .begin()
             .await
             .context("acquire DB connection")?;
-        let event = match database::settlements::get_settlement_without_auction(&mut ex)
+        let (hash, event) = match database::settlements::get_settlement_without_auction(&mut ex)
             .await
             .context("get_settlement_without_auction")?
         {
-            Some(event) => event,
+            Some(event) => (
+                domain::eth::TxId(H256(event.tx_hash.0)),
+                domain::eth::Event {
+                    block: (event.block_number as u64).into(),
+                    log_index: event.log_index as u64,
+                },
+            ),
             None => return Ok(false),
         };
 
-        let hash = H256(event.tx_hash.0);
         tracing::debug!("updating settlement details for tx {hash:?}");
 
-        let Ok(transaction) = self.eth.transaction(hash.into()).await else {
+        let Ok(transaction) = self.eth.transaction(hash).await else {
             tracing::warn!(?hash, "no tx found");
             return Ok(false);
         };
         let domain_separator = self.eth.contracts().settlement_domain_separator();
-
-        let (auction_id, auction_data) = match Self::recover_auction_id_from_calldata(
-            &mut ex,
-            &transaction,
-            &model::DomainSeparator(domain_separator.0),
+        let settlement = domain::settlement::Settlement::new(
+            transaction.clone(),
+            domain_separator,
+            &self.persistence,
         )
-        .await?
-        {
-            AuctionIdRecoveryStatus::InvalidCalldata => {
-                // To not get stuck on indexing the same transaction over and over again, we
-                // insert the default auction ID (0)
-                (Default::default(), None)
-            }
-            AuctionIdRecoveryStatus::DoNotAddAuctionData(auction_id) => (auction_id, None),
-            AuctionIdRecoveryStatus::AddAuctionData(auction_id, settlement) => (
-                auction_id,
-                Some(
-                    self.fetch_auction_data(settlement, auction_id, &transaction, &mut ex)
-                        .await?,
-                ),
-            ),
-        };
+        .await;
 
-        let update = SettlementUpdate {
-            block_number: event.block_number,
-            log_index: event.log_index,
-            auction_id,
-            auction_data,
+        let auction_id = match &settlement {
+            Ok(settlement) => settlement.auction_id(),
+            Err(domain::settlement::Error::Infra(err)) => {
+                return Err(anyhow!("{hash:?}, infra error {err}"))
+            }
+            Err(err) => {
+                tracing::warn!(?hash, ?err, "invalid settlement");
+                match domain::settlement::Solution::new(&transaction.input, domain_separator) {
```

Comment:
I don't think we should try to create a Solution again here. If the sole purpose is to extract the auction id, can we not expose this on the Settlement error directly?